Mathlib/RingTheory/TensorProduct/MvPolynomial.lean

```diff
@@ -0,0 +1,169 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.LinearAlgebra.DirectSum.Finsupp
+import Mathlib.Algebra.MvPolynomial.Basic
+import Mathlib.RingTheory.TensorProduct.Basic
+import Mathlib.Algebra.MvPolynomial.Equiv
+/-!
+
+# Tensor Product of (multivariate) polynomial rings
+
+* `MvPolynomial.rTensor`, `MvPolynomial.scalarRTensor`:  the tensor product of
+  a polynomial algebra by a module is linearly equivalent
+  to a Finsupp of a tensor product
+* `MvPolynomial.rTensorAlgHom`, the algebra morphism from the tensor product
+  of a polynomial algebra by an algebra to a polynomial algebra
+* `MvPolynomial.rTensorAlgEquiv`, `MvPolynomial.scalarRTensorAlgEquiv`,
+  the tensor product of a polynomial algebra by an algebra
+  is algebraically equivalent to a polynomial algebra
+
+## TODO :
+* `MvPolynomial.rTensor` could be phrased in terms of `AddMonoidAlgebra`, and
+  `MvPolynomial.rTensor` then has `smul` by the polynomial algebra.
+* `MvPolynomial.rTensorAlgHom` and `MvPolynomial.scalarRTensorAlgEquiv`
+  are morphisms for the algebra structure by `MvPolynomial σ R`.
+-/
+
+
+universe u v w
+
+noncomputable section
+
+namespace MvPolynomial
+
+open DirectSum TensorProduct
+
+open Set LinearMap Submodule
+
+variable {R : Type u} {M : Type v} {N : Type w}
+  [CommSemiring R] [AddCommMonoid M] [Module R M]
+
+variable {σ : Type*} [DecidableEq σ]
+
+variable {S : Type*} [CommSemiring S] [Algebra R S]
+
+section Module
+
+variable   [AddCommMonoid N] [Module R N]
+
+/-- The tensor product of a polynomial ring by a module is
+  linearly equivalent to a Finsupp of a tensor product -/
+noncomputable def rTensor :
+    MvPolynomial σ S ⊗[R] N ≃ₗ[S] (σ →₀ ℕ) →₀ (S ⊗[R] N) :=
+  TensorProduct.finsuppLeft' _ _ _ _ _
+
+lemma rTensor_apply_tmul (p : MvPolynomial σ S) (n : N) :
+    rTensor (p ⊗ₜ[R] n) = p.sum (fun i m ↦ Finsupp.single i (m ⊗ₜ[R] n)) :=
+  TensorProduct.finsuppLeft_apply_tmul p n
+
+lemma rTensor_apply_tmul_apply (p : MvPolynomial σ S) (n : N) (d : σ →₀ ℕ) :
+    rTensor (p ⊗ₜ[R] n) d = (coeff d p) ⊗ₜ[R] n :=
+  TensorProduct.finsuppLeft_apply_tmul_apply p n d
```

Comment:
Should this be a simp lemma?
Also I think there should be `rTensor_apply_monomial_tmul` and `rTensor_apply_X_tmul` as simp lemmas as well.

Reply:
similarly for `scalarRTensor`.

Reply:
I'll do that! thanks for the suggestion.
(Generally speaking, I have little idea of what lemmas should be made simp or not.)